File: src/cappuccino/ide/intellij/plugin/fixes/ObjJAddSemiColonIntention.java
package cappuccino.ide.intellij.plugin.fixes;

import cappuccino.ide.intellij.plugin.inspections.ObjJInspectionProvider;
import cappuccino.ide.intellij.plugin.psi.interfaces.ObjJNeedsSemiColon;
import cappuccino.ide.intellij.plugin.utils.EditorUtil;
import com.intellij.codeInsight.intention.impl.BaseIntentionAction;
import com.intellij.openapi.editor.Editor;
import com.intellij.openapi.project.Project;
import com.intellij.psi.PsiFile;
import com.intellij.util.IncorrectOperationException;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;

public class ObjJAddSemiColonIntention extends BaseIntentionAction {
    private final ObjJNeedsSemiColon element;

    public ObjJAddSemiColonIntention(final ObjJNeedsSemiColon element) {
        this.element = element;
    }

    @NotNull
    @Override
    public String getText() {
        return "Insert missing semi-colon";
    }

    @Nls
    @NotNull
    @Override
    public String getFamilyName() {
        return ObjJInspectionProvider.GROUP_DISPLAY_NAME;
    }

    @Override
    public boolean isAvailable(
            @NotNull
                    Project project, Editor editor, PsiFile psiFile) {
        return !this.element.getLastChild().getText().equals(";");
    }

    @Override
    public void invoke(
            @NotNull
                    Project project, Editor editor, PsiFile psiFile) throws IncorrectOperationException {
        EditorUtil.INSTANCE.insertText(editor, ";", element.getTextRange().getEndOffset(), true);
    }
}

File: src/main/java/cappuccino/ide/intellij/plugin/fixes/ObjJAddSemiColonIntention.kt
package cappuccino.ide.intellij.plugin.fixes

import cappuccino.ide.intellij.plugin.inspections.ObjJInspectionProvider
import cappuccino.ide.intellij.plugin.lang.ObjJBundle.message
import cappuccino.ide.intellij.plugin.psi.interfaces.ObjJNeedsSemiColon
import cappuccino.ide.intellij.plugin.utils.EditorUtil.insertText
import cappuccino.ide.intellij.plugin.utils.editor
import com.intellij.codeInsight.intention.IntentionAction
import com.intellij.codeInspection.LocalQuickFix
import com.intellij.codeInspection.ProblemDescriptor
import com.intellij.openapi.editor.Editor
import com.intellij.openapi.project.Project
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiFile
import com.intellij.psi.SmartPointerManager
import com.intellij.psi.SmartPsiElementPointer

class ObjJAddSemiColonIntention(element: ObjJNeedsSemiColon) : IntentionAction, LocalQuickFix {
    private val element: SmartPsiElementPointer<ObjJNeedsSemiColon>
    override fun startInWriteAction(): Boolean = true
    override fun getText(): String = message("objective-j.intentions.add-semi-colon.prompt")
    override fun getName(): String = message("objective-j.intentions.add-semi-colon.prompt")
    override fun getFamilyName(): String = ObjJInspectionProvider.GROUP_DISPLAY_NAME

    override fun isAvailable(
        project: Project, editor: Editor?, psiFile: PsiFile?
    ): Boolean {
        val element = element.element ?: return false
        return element.lastChild.text != ";"
    }

    override fun applyFix(project: Project, descriptor: ProblemDescriptor) {
        val element = descriptor.psiElement
            ?: return
        val editor = element.editor
            ?: return
        invoke(editor, element)
    }

    override fun invoke(
        project: Project, editor: Editor, psiFile: PsiFile?
    ) {
        val element = element.element ?: return
        invoke(editor, element)
    }

    private fun invoke(editor: Editor, element: PsiElement) {
        insertText(editor, ";", element.textRange.endOffset, true)
    }

    init {
        this.element = SmartPointerManager.createPointer(element)
    }
}